Properly retrieve folder names from html pages of all mirrors

mite-user opened the pull request:
When https://download.qt.io is not available, `aqt list-*` command falls back to a random mirror.
Different mirrors provide different html layouts.

Mirrors with `<table>` similar enough to be properly handled by the current script:
https://download.qt.io/
https://ftp.jaist.ac.jp/pub/qtproject/
https://qt-mirror.dannhauer.de/
https://master.qt.io/

Mirrors with `<table>` which isn't properly handled by the current script:
https://mirrors.aliyun.com/qt/
https://mirrors.sjtug.sjtu.edu.cn/qt/
https://mirrors.dotsrc.org/qtproject/

Mirrors with a list of `<a>` elements inside `<pre>`, no `<table>`:
https://qtproject.mirror.liquidtelecom.com/
https://mirrors.ustc.edu.cn/qtproject/
https://ftp.yz.yamagata-u.ac.jp/pub/qtproject/
https://ftp.fau.de/qtproject/
https://mirror.netcologne.de/qtproject/
https://www.nic.funet.fi/pub/mirrors/download.qt-project.org/
https://mirrors.ukfast.co.uk/sites/qt.io/
https://qt.mirror.constant.com/

Mirrors with a list of `<a>` elements inside `<pre>`, with `<table>` which doesn't contain folders:
https://ftp2.nluug.nl/languages/qt/
https://ftp1.nluug.nl/languages/qt/


If the mirror has similar enough html layout, the program works and prints the requested info.

If there is no `<table>` element,
v2.1.0 fails with the error described in #519 ;
the master branch prints
```
Failed to retrieve the expected HTML page at https://download.qt.io/online/qtsdkrepository/linux_x64/desktop/
```
This is an unrelated bug. The program should be printing the fallback url, not baseurl.

If there is a `<table>` element, the program prints
```
No versions available for this request.
```
or
```
No tools available for this request.
```



## Solution.
Get folder names from value of `href` attribute of `<a>` elements.
Folder links point to a page one level deeper relative to the current page,
i.e. the value should look like `./foo` or `foo`,
NOT `https://foo`, `../foo`, `..`, `/foo`, `?foo`.



## How to test.
```
git clone https://github.com/mite-user/aqtinstall.git
echo '__version__ = "handle-html-test"' > aqtinstall/aqt/version.py
```
Open `aqtinstall/aqt/settings.ini`,
set `baseurl:` to `http://localhost`,
remove all mirrors expect the one to be tested from `fallbacks:`.
```
cd aqtinstall
python3 -m aqt list-qt linux desktop
```



This commit might fix #519 .

ddalcino commented on aqt/metadata.py:
```diff
@@ -589,20 +589,26 @@ def fetch_http(rest_of_url: str, is_check_hash: bool = True) -> str:
             try:
-                return tr.find_all("td")[1].a.contents[0].rstrip("/")
+                url: str = link.get("href")
+                url = url.split("?")[0].rstrip("/")
```
Have you looked at `urllib.parse`? If you import `urlparse` from `urllib.parse`, you can do this:
```suggestion
                url = urlparse(url).path.rstrip("/")
```
... and get the benefit of a thoughtfully-designed URL parsing library that will take out anything else that doesn't belong in the path (for instance, `#some_id`).

I'm not sure we really need to use `urlparse`, but I think it's worth asking.

ddalcino commented on aqt/metadata.py:
```diff
@@ -589,20 +589,26 @@ def fetch_http(rest_of_url: str, is_check_hash: bool = True) -> str:
-                content: str = table_row_to_folder(row)
-                if not content or content == "Parent Directory":
+            for link in soup.find_all("a"):
+                folder: str = link_to_folder(link)
```
Thank you for renaming this; its much easier to read now.

ddalcino reviewed (changes requested):
This looks good, and it's a welcome improvement! Thank you.

The only real suggestion I have is that you add a test or two that the existing implementation cannot pass. In your discussion of the problem, you mentioned several different cases where the mirror HTML pages have a different structure than the pages at download.qt.io. I would like to see tests that run `iterate_folders` or `get_versions_extensions` on a couple of examples of these differently-structured HTML files, and expect the correct set of results. This will help us to prevent regressions in the future.

mite-user commented:
### Updated the original commit
1. Using `urllib.parse.urlparse`.
2. Handling `./` at the beginning and `/` at the end with `posixpath.normpath`.
3. Checking for one and two dots only. Apparently three and more dots point one level deeper.



### Second commit
Added test html files for `test_list_versions_tools()` in `tests/test_list.py`.



### Notes
Absence of `<table>` element no longer triggers the error added in #518 . There should be some other way to check if html contents are unexpected. `test_list_bad_html()` now always fails.

`list-tool` doesn't sort the resulting list. So the order depends on what is provided by html. Looks like all or almost all mirrors provide sorted links by default. They are unsorted at https://download.qt.io .

ddalcino commented on aqt/metadata.py:
```diff
@@ -24,6 +24,7 @@
 import re
 import secrets as random
 import shutil
+from urllib.parse import urlparse
```
This part of the 'check' action: https://github.com/miurahr/aqtinstall/runs/6331277720?check_suite_focus=true#step:7:17 failed because the imports are no longer sorted. You can fix that with isort.

If this helps, I always run `isort aqt tests && black aqt tests` before I commit anything that's going into this codebase, just to avoid any code formatting or import sorting issues. I think you can fix them with `flake8` as well.

miurahr commented on aqt/metadata.py:
```diff
@@ -26,6 +26,7 @@
 from logging import getLogger
 from typing import Callable, Dict, Generator, Iterable, Iterator, List, Optional, Tuple, Union
+from urllib.parse import ParseResult, urlparse
 from xml.etree.ElementTree import Element
```
We have already use defusedxml as dependency. Please feel free to use it.

mite-user replied:
I think it's complaining about `from xml.etree.ElementTree import Element`. It's irrelevant to this pull request.
Not an issue, apparently.
https://github.com/miurahr/aqtinstall/pull/498#discussion_r824344337

mite-user commented:
### Changed `link_to_folder()`.

Removed `try:` `except (AttributeError, IndexError):`.
With the new code, exceptions are unexpected and shouldn't be ignored.

Changed `link.get("href")` to `link.get("href", default="")`.
Without `default=""`, it will return `None` instead of empty `str` if there is no `href` in `<a>`.

ddalcino commented:
> Absence of `<table>` element no longer triggers the error added in #518 . There should be some other way to check if html contents are unexpected. `test_list_bad_html()` now always fails. 

For the record, I’d like to say that I support removing the `test_list_bad_html()` test that now fails. I don't think that it really adds all that much. 

I wrote that test under the assumption that #517 was caused by being in a coffee shop, or by some http error. Given the evidence that @mite-user presents here, l think that it's much more likely that a different html structure in a mirror site was the cause. The requests library should be able to detect serious HTTP errors that would result in malformed html, and I don't think you get a 200 status code when a router asks you to sign in.

ddalcino commented:
@miurahr Any further thoughts on this? I still think we should remove `test_list_bad_html()` and merge this. I don't think there's any obvious way to merge this without removing that test, and I don't think the test is useful.

miurahr commented:
@ddalcino 
>  I still think we should remove test_list_bad_html() and merge this. 

I think it is OK to remove and merge this.

mite-user commented:
Should I add a commit to this pull request to remove `test_list_bad_html()`?

ddalcino commented:
> Should I add a commit to this pull request to remove `test_list_bad_html()`?

IMHO, yes, you should add a commit that removes the failing test. I don’t think we need a separate PR for this. If we did, I think we would have heard from @miurahr by now.

This PR effectively replaces #518, which is where `test_list_bad_html()` comes from. Speaking as the author of that test, I think the right way to replace it is to just remove it.

ddalcino reviewed (approved):
All tests are passing, and the important parts of this code have test coverage. It looks like `aqt/metadata.py:611-613` is no longer covered, but I don't think it matters. IMHO this is ready to merge!
